R/input-date.R

```diff
@@ -124,19 +130,98 @@ dateInput <- function(inputId, label, value = NULL, min = NULL, max = NULL,
+    scssDir <- system.file(package = "shiny", "www", "shared", "datepicker", "scss")
+    tmpDir <- tempfile("datepicker-scss")
+    dir.create(tmpDir)
+    outFile <- file.path(tmpDir, "custom.css")
+    sassFunc <- if (useBsTheme()) bootstraplib::bootstrap_sass else sass::sass
+    defaults <- list(bsThemeDateDefaults(), sassVars)
+    sassFunc(
+      list(
+        defaults,
+        sass::sass_file(file.path(scssDir, "build3.scss"))
+      ),
+      output = outFile
+    )
+    # Hash the defaults so that the html dependency from one date-picker won't stomp another
+    cssDependency <- htmlDependency(
+      paste0("datepicker-datepicker-css-", digest::digest(defaults, "xxhash64")),
+      version = datePickerVersion,
+      src = dirname(outFile),
+      stylesheet = basename(outFile)
+    )
+  } else {
+    cssDependency <- htmlDependency(
+      "bootstrap-datepicker-css",
+      version = datePickerVersion,
+      src = c(href = "shared/datepicker"),
+      stylesheet = "css/bootstrap-datepicker3.min.css",
+    )
+  }
+
+  list(
+    cssDependency,
+    htmlDependency(
+      "bootstrap-datepicker-js",
+      version = datePickerVersion,
+      c(href = "shared/datepicker"),
+      script = "js/bootstrap-datepicker.min.js",
+      # Need to enable noConflict mode. See #1346.
+      head = "<script>(function() {
+        var datepicker = $.fn.datepicker.noConflict();
+        $.fn.bsDatepicker = datepicker;
+      })();
+     </script>"
+    )
+  )
+}
+
+
+bsThemeDateDefaults <- function() {
+  if (!useBsTheme()) {
+    return(list())
+  }
+
+  version <- bootstraplib::theme_version()
+  if ("3" %in% version) {
+    return(list(
+      "@function color-contrast($color, $dark: $gray-darker, $light: $text-color) {
+        @return if(
+          red($color) * 0.299 + green($color) * 0.587 + blue($color) * 0.114 > 150,
+          $dark, $light
+        );
+      }",
+      # No !default flag because these are variables already defined by Bootstrap
+      list(
+        "btn-primary-color" = "color-contrast($btn-primary-bg)",
+        "btn-primary-color-hover" = "color-contrast($gray-lighter)",
+        "state-info-bg" = "mix($body-bg, $brand-primary, 20%)",
+        "dropdown-bg" = "$body-bg",
+        "dropdown-border" = "mix($text-color, $body-bg, 15%)"
+      )
+    ))
+  }
+
+  list(
+    "gray" = "gray('300')",
+    "gray-light" = "gray('500')",
+    "gray-lighter" = "gray('900')",
+    "brand-primary" = "mix($black, theme-color('primary'), 6.5%)",
+    "btn-primary-color" = "color-yiq($brand-primary)",
+    "btn-primary-color-hover" = "color-yiq($gray-lighter)",
+    "btn-primary-bg" = "$brand-primary",
+    "btn-primary-border" = "darken($btn-primary-bg, 5%)",
+    "btn-link-disabled-color" = "$gray-light",
+    "state-info-bg" = "mix($white, $brand-primary, 20%)",
+    "line-height-base" = "20/14",
+    "border-radius-base" = "4px",
+    "dropdown-bg" = "$white",
+    "dropdown-border" = "mix($black, $white, 15%)"
```

Comment:
Turns out that Bootswatch doesn't set `$white`/`$black` to true bg/fg colors, so I think we should replace `$white` with `$body-bg` and `$black` with `$body-color`

Reply:
(The other theming PRs suffer from the same issue....maybe we should "shim" bootswatch sass variables in bootstraplib so that they're structured more like `bs_theme_base_colors()`'s model